pulsar-client/src/main/java/org/apache/pulsar/client/impl/PersistentAcknowledgmentsGroupingTracker.java

```diff
@@ -477,13 +477,16 @@ public void flush() {
     private void flushAsync(ClientCnx cnx) {
         boolean shouldFlush = false;
         if (cumulativeAckFlushRequired) {
-            newMessageAckCommandAndWrite(cnx, consumer.consumerId, lastCumulativeAck.messageId.ledgerId,
-                    lastCumulativeAck.messageId.getEntryId(), lastCumulativeAck.bitSetRecyclable,
-                    AckType.Cumulative, null, Collections.emptyMap(), false,
-                    this.currentCumulativeAckFuture, null);
-            this.consumer.unAckedChunkedMessageIdSequenceMap.remove(lastCumulativeAck.messageId);
-            shouldFlush = true;
-            cumulativeAckFlushRequired = false;
+            final MessageIdImpl messageIdOfLastAck = lastCumulativeAck.messageId;
```

Comment:
Can you explain why we enter this code block when `lastCumulativeAck.messageId` is null? It seems like `cumulativeAckFlushRequired` should be false. I'm fine with adding a null check, but it seems like there could be another issue.

Reply:
The previous logic showed, stat were updated only when this.consumer.unAckedChunkedMessageIdSequenceMap.remove successfully.

Reply:
@gaozhangmin When reading the code, it seems that messageId should never be null. As @michaeljmarshall pointed out, there could be another issue hiding here. 
It might be a thread safety issue which could cause other problems. @gaozhangmin  Did you check how messageId could become null?

Reply:
@lhotari  No, i didn't check the reason. much same as this issue: #11607

Reply:
PR #11607 fixes another case where lastCumulativeAck.messageId was null. @BewareMyPower What are your thoughts about the reason why the field is null in the first place? Could there be a thread safety issue that won't go away by adding null checks?

Reply:
@gaozhangmin do you have a chance to analyse the reason? My concern is that adding a null check might just silently ignore a potential thread safety issue which might lead to inconsistency behavior.

Reply:
@lhotari As I've explained in my PR, I think there is a possible thread safety problem like

```java
 if (lastCumulativeAck.messageId == null) { // 1. messageId is not null
     return false; 
 } 
 // 2. messageId was modified to null in another thread
 if (messageId.compareTo(lastCumulativeAck.messageId) <= 0) { // 3. messageId is null now
```


I think the root cause is the design of `LastCumulativeAck` class. It's not well encapsulated. Though its fields are all private, the class is an inner class so the outer class can access the members directly. And we can see the direct access in many places, which makes it hard to analyze whether all the accesses are thread safe.

Reply:
> I think the root cause is the design of LastCumulativeAck class
@BewareMyPower  I also think that LastCumulativeAck class is the source of problems. Using the Netty Recycler introduces the thread safety issue. Using an immutable class design for LastCumulativeAck and removing the use of Netty recycler would be something that I'd recommend for fixing the issue.

Reply:
Sure, I will  implement  immutable class design.
What's about this pr. should we close it? @lhotari

Reply:
@gaozhangmin @lhotari I picked up this issue today. After looking deeper into this issue, I found another bug and it would be better to synchronize the update operations of `LastCumulativeAck`. See https://github.com/apache/pulsar/pull/16072, PTAL when you have time.